jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java

```diff
@@ -115,28 +115,67 @@ protected void doStop() throws Exception
     {
         if (LOG.isDebugEnabled())
             LOG.debug("Stopping {}", this);
-        CloseEndPoints close_endps = new CloseEndPoints();
-        submit(close_endps);
-        close_endps.await(getStopTimeout());
-        CloseSelector close_selector = new CloseSelector();
-        submit(close_selector);
-        close_selector.await(getStopTimeout());
-
-        super.doStop();
         
+        try
+        {
+            CloseAction close = new CloseAction();
+            submit(close);
+            if (getStopTimeout()>0)
+                close.await(getStopTimeout());
+        }
+        finally
+        {
+            try
+            {
+                super.doStop();
+            }
+            finally
+            {
+                try (Locker.Lock lock = _locker.lock())
+                {
+                    if (_selector!=null && _selector.isOpen())
+                        _selector.wakeup();
```

Comment:
If above you return a boolean, then you don't need these 2 lines above.
If the boolean is true, you know the selector has been closed, otherwise there is no need to wake it up (it has already been woken up by the submission of the `CloseAction`)

Reply:
the previous submit is before the call to super.doStop(), so it may wakeup, loop and then select again before we grab the lock at 134.   Once we have the lock we know the state of the ManagedSelector is `!isRunning()`, so another wakeup is needed to make sure the selecting thread sees that.